MINOR: Use time constant algorithms when comparing passwords or keys

rhauch opened the pull request.

rhauch committed:
MINOR: Use time constant algorithms when comparing passwords or keys

omkreddy commented on clients/src/main/java/org/apache/kafka/common/security/plain/internals/PlainServerCallbackHandler.java:
```diff
@@ -65,7 +65,7 @@ protected boolean authenticate(String username, char[] password) throws IOExcept
                     JAAS_USER_PREFIX + username,
                     PlainLoginModule.class.getName());
-            return expectedPassword != null && Arrays.equals(password, expectedPassword.toCharArray());
+            return expectedPassword != null && Utils.isEqual(password, expectedPassword.toCharArray());
```
How about using `MessageDigest.isEqual` method here also:

```
MessageDigest.isEqual(new String(password).getBytes(StandardCharsets.UTF_8), expectedPassword.getBytes(StandardCharsets.UTF_8)`
```

cc @rajinisivaram

rajinisivaram replied:
Yes, it should be safe to convert to bytes and compare using MessageDigest.isEqual since SASL/PLAIN uses UTF8. Means more object creation, but that shouldn't be an issue. Utils.isEqual() may be worth adding if we think we may use it in other places as well in future for constant time array comparisons (in which case, we can make it generic).

ijuma replied:
Hmm, if we convert arrays to bytes, we need to be careful. If the arrays have different sizes, then the operation is not constant time.

ijuma replied:
Hmm, if we convert arrays to bytes, we need to be careful. If the arrays have different sizes, then the operation is not constant time.

rajinisivaram replied:
Good point. Btw, this particular class (the default SASL/PLAIN server-side callback handler) is not recommended for production use, so perhaps not that critical.

mimaison commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -331,6 +331,41 @@ public static ByteBuffer wrapNullable(byte[] array) {
+        }
+
+        // time-constant comparison that always compares all characters in first array
+        boolean matches = first.length == second.length;
```
Can we return `false` directly if the lengths don't match? That would also remove the need for `int j = i < second.length ? i : 0;` when lengths match

mimaison commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -331,6 +331,41 @@ public static ByteBuffer wrapNullable(byte[] array) {
+        for (int i = 0; i < first.length; ++i) {
+            int j = i < second.length ? i : 0;
+            if (first[i] != second[j]) {
+                matches = false;
```
We can return `false` on the first mismatch no need to check the rest of the arrays.

rajinisivaram replied:
I believe the goal is to use constant-time comparison to prevent timing attacks, hence the walk through the arrays.

mimaison replied:
I see, my bad. What about renaming the method to make it explicit, something like `isEqualConstantTime()`?

mimaison reviewed (dismissed):
Thanks, I've left a couple of comments on `isEqual()`

rhauch committed:
MINOR: Rename constant time char array equality utility method

rhauch commented:
Thanks for the review, folks. @omkreddy, I renamed the new utility method as you suggested, and updated the JavaDocs as well. Hopefully the purpose and behavior is more clear.

rhauch commented:
@mimaison, @omkreddy, @rajinisivaram, @ijuma: would some of you be able to review these changes? Thanks!

omkreddy reviewed (approved):
@rhauch Thanks for the PR. LGTM.

rhauch commented:
The failing tests are unrelated to the changes.
